packages/core/src/Room.ts

```diff
@@ -583,9 +583,24 @@ export abstract class Room<State extends object= any, Metadata= any> {
         if (this.onJoin) {
           await this.onJoin(client, options, client.auth);
         }
+
+        client.ref.removeListener('close', client.ref['onleave']);
+
+        // client left during `onJoin`, call _onLeave immediately.
+        if (client.state === ClientState.LEAVING) {
+          await this._onLeave(client, Protocol.WS_CLOSE_GOING_AWAY);
```

Comment:
by calling _onLeave here, the _onLeave function will emit `leave` event to matchmaker before the 'join'
The net should be correct, since the 'join' is still sent below, but emitting these out of order seems potentially problematic.

Reply:
Well observed! It turns out it's safe to move the `'join'` event before this call, as it is already being explicitly ignored if `!previousReconnectionToken`. So that's the only code path the `'join'` event could happen 🙌